client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -120,10 +125,12 @@ public synchronized void setHosts(HttpHost... hosts) {
         if (hosts == null || hosts.length == 0) {
             throw new IllegalArgumentException("hosts must not be null nor empty");
         }
+        authCache.clear();
         Set<HttpHost> httpHosts = new HashSet<>();
         for (HttpHost host : hosts) {
             Objects.requireNonNull(host, "host cannot be null");
             httpHosts.add(host);
+            authCache.put(host, new BasicScheme());
```

Comment:
volatile reads are cheap on x86; IIRC mapping to a no-op. I'd prefer volatile; but there is still the issue that there will be a small amount of time when hosts and the auth cache are not in sync (between setting both).